# Chapter 3 -  Metatranscriptomic data preparation

## 3.1 - Load data

### Metatranscriptomic gene expression counts

```{r load_expr_counts, warning = FALSE, comments = "", message = FALSE, eval = FALSE}
load("data/distilled_caecum_1.RData")
load("data/distilled_caecum_2.RData")
load("data/distilled_caecum_3.RData")
load("data/distilled_caecum_4.RData")
load("data/distilled_caecum_5.RData")
load("data/distilled_caecum_6.RData")
load("data/distilled_caecum_7.RData")
load("data/distilled_caecum_8.RData")

expr_counts_raw <- c(distilled_expression_caecum_1,distilled_expression_caecum_2,distilled_expression_caecum_3,distilled_expression_caecum_4,distilled_expression_caecum_5,distilled_expression_caecum_6,distilled_expression_caecum_7,distilled_expression_caecum_8)

rm(distilled_expression_caecum_1,distilled_expression_caecum_2,distilled_expression_caecum_3,distilled_expression_caecum_4,distilled_expression_caecum_5,distilled_expression_caecum_6,distilled_expression_caecum_7,distilled_expression_caecum_8)

# Change MAG IDs for a standardised code 
ena_to_mag_id <- read_tsv("data/ena_to_mag_id.tsv")

mag_name <- names(expr_counts_raw)

names(expr_counts_raw) <- ena_to_mag_id$mag_id[match(mag_name, ena_to_mag_id$mag_name)]

# Correct animal codes
expr_counts <- lapply(expr_counts_raw, function(x) {
  rownames(x) <- gsub("F1a", "", rownames(x))
  x
})
```

### Metadata of chickens sequenced for MT

```{r load_metadata, warning = FALSE, comments = "", message = FALSE, eval = FALSE}
animal_codes_mt <- rownames(expr_counts$cmag_001) %>% 
  gsub("F1a", "", .)

chicken_metadata_mt <- 
  read_tsv("data/metadata.tsv") %>%
  mutate(sampling_time = factor(sampling_time, levels = c('7',
                                                          '21',
                                                          '35'))) %>% 
  filter(animal_code %in% animal_codes_mt)
```

## 3.2 - Save working objects

```{r save_mt_objects, warning = FALSE, comments = "", message = FALSE, eval = FALSE}
save(expr_counts, chicken_metadata_mt, file = "data/data_mt.Rdata")
```

### Clean environment

```{r clean, comment = "", message = FALSE, warning = FALSE}
rm(list = ls())
```

