# Bacterial community composition

## Load MG data

```{r load_mg_data_beta_div}
load("data/data_mg.Rdata")
```

## Hill numbers - beta diversity

```{r beta_div, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
# Neutral
beta_q1n <- 
  mag_weighted %>% 
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  dplyr::select_if(~!all(. == 0)) %>%
  hillpair(., q = 1, metric = "S", out = "pair")

# Phylogenetic
beta_q1p <- 
  mag_weighted %>% 
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  dplyr::select_if(~!all(. == 0)) %>%
  hillpair(., q = 1, tree = genome_tree, metric = "S", out = "pair")

# Functional KEGG
dist_kegg <-
  genome_kegg_paths %>%
  column_to_rownames(var = "genome") %>%
  traits2dist(., method = "gower")

beta_q1f_kegg <-
  mag_weighted %>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  dplyr::select_if(~!all(. == 0)) %>%
  hillpair(., q = 1, dist = dist_kegg, metric = "S", out = "pair")
```

```{r save_beta, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
save(beta_q1n, 
     beta_q1p, 
     beta_q1f_gifts,
     file = "data/beta_div.Rdata")
```

### Compare individuals from same trial and pen between sampling days 

The resulting stats corresponds to Supplementary Table S3 in the manuscript.

```{r compare_ind, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
load("data/beta_div.Rdata")

metadata_2 <-
  chicken_metadata %>%
  dplyr::rename(second = 'animal_code',
                sampling_time_2 = 'sampling_time',
                pen_2 = 'pen',
                trial_2 = 'trial')

# Neutral 
betadiv_n_dis <-
  beta_q1n %>% 
  rename(animal_code = 'first') %>%
  left_join(chicken_metadata, by = 'animal_code') %>%
  left_join(metadata_2, by = 'second') %>%
  mutate(diff = case_when(
    sampling_time == '7' & sampling_time_2 == '21' ~ '7_21',
    sampling_time == '21' & sampling_time_2 == '35' ~ '21_35',
    sampling_time == '7' & sampling_time_2 == '35' ~ '7_35')) %>%
  mutate(diff_trial = case_when(trial == 'CA' & trial_2 == 'CA' ~ 'CA',
                                trial == 'CB' & trial_2 == 'CB' ~ 'CB')) %>%
  mutate(diff_pen = case_when(pen == pen_2  ~ 'same')) %>%
  drop_na(diff, diff_trial, diff_pen)

# Phylogenetic
betadiv_p_dis <-
  beta_q1p %>% 
  rename(animal_code = 'first') %>%
  left_join(chicken_metadata, by = 'animal_code') %>%
  left_join(metadata_2, by = 'second') %>%
  mutate(diff = case_when(
    sampling_time == '7' & sampling_time_2 == '21' ~ '7_21',
    sampling_time == '21' & sampling_time_2 == '35' ~ '21_35',
    sampling_time == '7' & sampling_time_2 == '35' ~ '7_35')) %>%
  mutate(diff_trial = case_when(trial == 'CA' & trial_2 == 'CA' ~ 'CA',
                                trial == 'CB' & trial_2 == 'CB' ~ 'CB')) %>%
  mutate(diff_pen = case_when(pen == pen_2  ~ 'same')) %>%
  drop_na(diff, diff_trial, diff_pen)

# Functional KEGG
betadiv_f_dis_kegg <-
  beta_q1f_kegg %>%
  rename(animal_code = 'first') %>%
  left_join(chicken_metadata, by = 'animal_code') %>%
  left_join(metadata_2, by = 'second') %>%
  mutate(diff = case_when(
    sampling_time == '7' & sampling_time_2 == '21' ~ '7_21',
    sampling_time == '21' & sampling_time_2 == '35' ~ '21_35',
    sampling_time == '7' & sampling_time_2 == '35' ~ '7_35')) %>%
  mutate(diff_trial = case_when(trial == 'CA' & trial_2 == 'CA' ~ 'CA',
                                trial == 'CB' & trial_2 == 'CB' ~ 'CB')) %>%
  mutate(diff_pen = case_when(pen == pen_2  ~ 'same')) %>%
  drop_na(diff, diff_trial, diff_pen)

# Stats
betadiv_n_dis %>%
  group_by(diff) %>%
  summarise(mean = mean(S), sd = sd(S))

betadiv_p_dis %>%
  group_by(diff) %>%
  summarise(mean = mean(S), sd = sd(S))

betadiv_f_dis_kegg %>%
  group_by(diff) %>%
  summarise(mean = mean(S), sd = sd(S))
```

### Plot difference between sampling points for beta diversity 

Not included in the manuscript

```{r plot_beta_diff, comment="", message = FALSE, warning = FALSE, fig.height = 4, fig.width = 5, fig.fullwidth = TRUE, eval = FALSE}
# Neutral
p_n <- 
  betadiv_n_dis %>%
  filter(!diff == '7_35') %>%
  mutate(diff = factor(diff, levels = c('7_21', '21_35'))) %>%
  ggplot(aes(x = diff, y = S, fill = diff)) +
  geom_boxplot() +
  ylim(0,0.5) +
  theme_minimal() +
  theme(legend.position = 'none') +
  ggtitle('Neutral beta div')

# Phylogenetic
p_p <- 
  betadiv_p_dis %>%
  filter(!diff == '7_35') %>%
  mutate(diff = factor(diff, levels = c('7_21', '21_35'))) %>%
  ggplot(aes(x = diff, y = S, fill = diff)) +
  geom_boxplot() +
  ylim(0,0.5) +
  theme_minimal() +
  theme(legend.position = 'none') +
  ggtitle('Phylogenetic beta div')

# Functional KEGG
p_f_kegg <-
  betadiv_f_dis_kegg %>%
  filter(!diff == '7_35') %>%
  mutate(diff = factor(diff, levels = c('7_21', '21_35'))) %>%
  ggplot(aes(x = diff, y = S, fill = diff)) +
  geom_boxplot() +
  ylim(0,0.5) +
  theme_minimal() +
  theme(legend.position = 'none') +
  ggtitle('Functional KEGG beta div')


grid.arrange(p_n, p_p, p_f_kegg, ncol = 3)
```

## Effect of design in microbiome composition

### Permanova values for different metrics of beta diversity 

The resulting stats correspond to Supplementary Table S3 in the manuscript.

```{r permanova_betadiv, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
# Dissimilarity tables
perm <- how(nperm = 999)
setBlocks(perm) <- with(chicken_metadata, pen)

# Neutral
b1n_dis_table <- 
  beta_q1n %>%
  bind_rows(beta_q1n %>% rename(first = second, second = first)) %>% 
  pivot_wider(names_from = second, values_from = S) %>%
  column_to_rownames('first') %>%
  mutate(across(everything(), ~replace(., is.na(.), 0))) %>% 
  as.dist()

adonis2(b1n_dis_table ~ trial + sex + breed + treatment * age,
        permutations = perm,
        data = chicken_metadata)

# Phylogenetic
b1p_dis_table <- 
  beta_q1p %>%
  bind_rows(beta_q1n %>% rename(first = second, second = first)) %>% 
  pivot_wider(names_from = second, values_from = S) %>%
  column_to_rownames('first') %>%
  mutate(across(everything(), ~replace(., is.na(.), 0))) %>% 
  as.dist()

adonis2(b1p_dis_table ~ trial + sex + breed + treatment * age,
        permutations = perm,
        data = chicken_metadata)

# Functional KEGG
b1f_dis_table_kegg <-
  beta_q1f_kegg %>%
  bind_rows(beta_q1n %>% rename(first = second, second = first)) %>%
  pivot_wider(names_from = second, values_from = S) %>%
  column_to_rownames('first') %>%
  mutate(across(everything(), ~replace(., is.na(.), 0))) %>%
  as.dist()

adonis2(b1f_dis_table_kegg ~ trial + sex + breed + treatment * age,
        permutations = perm,
        data = chicken_metadata)
```

### Community composition

```{r composition, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
perm <- how(nperm = 999)
setBlocks(perm) <- with(chicken_metadata, pen)

adonis2(hel ~ trial * sampling_time + 
              breed * sampling_time +
              sex * sampling_time +
              treatment * sampling_time,
        permutations = perm,
        data = chicken_metadata)

adonis2(hel ~ trial * sampling_time + breed + sex + treatment,
        permutations = perm,
        data = chicken_metadata)
```

## Community composition development

### Distance-based RDA 

The resulting plot corresponds to Figure 1f in the manuscript.

```{r distance_rda, fig.height = 4, fig.width = 7, fig.fullwidth = TRUE, eval = FALSE}
set.seed(4)
hel_bray <- vegdist(hel, method = 'bray')
hel_bray_sqrt <- sqrt(hel_bray)

pcoa <- cmdscale(hel_bray_sqrt, k = nrow(hel) - 1, eig = TRUE)
pcoa_scores <- pcoa$points

db_rda <- rda(pcoa_scores ~ sampling_time, data = chicken_metadata)

# Stats
anova(db_rda, by = 'term')
summary(db_rda)
RsquareAdj(db_rda)

db_rda_scores <-
  data.frame(
    scores(db_rda, display = 'wa'),
    pen = chicken_metadata$pen,
    time = chicken_metadata$sampling_time
    )

db_rda_scores %>%
  ggplot(aes(x = RDA1, y = RDA2, colour = time)) +
  geom_point() +
  scale_color_manual(values = c('#e6a024', '#cc6777', '#5bb4e5')) +
  theme_bw() +
  theme(legend.position = 'none')
```

```{r clean_6, comment = "", message = FALSE, warning = FALSE, eval = FALSE}
rm(list = ls())
```
